### ISSUE TITLE
Reduce Prometheus Data Retention Duration

### ISSUE THREAD

**Issue by ihcsim** [linkerd/linkerd2#3997]
Following a [recent slack report](https://linkerd.slack.com/archives/C89RTCWJF/p1580228200108300) on a Prometheus OOMKilled incident, we want to revisit the effect of reducing the `storage.tsdb.retention.time` duration from the current default of 6 hours to 1 hour. The previous investigation, per https://github.com/linkerd/linkerd2/issues/3258#issuecomment-526377964, showed that it did help to reduce the memory footprint. The obvious tradeoff is that there will be less metrics data available to the user.

It was also reported that reducing the `storage.tsdb.min-block-duration` to 30 mins and `storage.tsdb.max-block-duration` to 1 hour, from the default 2 hours, reduced Prometheus' memory usage by less than half. Note that these are [hidden flags](https://github.com/prometheus/prometheus/blob/9c67fce6e0123e3acc85c6ec4cc89e79cb2010d8/cmd/prometheus/main.go#L186-L188), marked for "testing only". We may or may not want to rely on them, but it's still worth investigating.

Open Questions:

* Shall we introduce make these settings configurable via CLI flags, config file (see https://github.com/linkerd/linkerd2/issues/3870), K8s configmap etc.?

**Comment by sannimichaelse** [comment]
Hi @grampelberg , Please can you assign this task to me, I would like to work on it. 

Also I will need more clarity on the question

**Comment by grampelberg** [comment]
@sannimichaelse take a look at the issue, there's quite a bit of detail there. You'll be modifying the prometheus configuration.

**Comment by ihcsim** [comment]
> Also I will need more clarity on the problem

@sannimichaelse What is your question?

**Comment by sannimichaelse** [comment]
I think it's clear now, will have to change the retention time on the prometheous configuration from 6h to 1h.

**Comment by ihcsim** [comment]
That's correct. Ideally, there should be a before/after comparison on the Linkerd Prometheus' memory and CPU usage, to justify this change. You can take a look at how I did it previously at https://github.com/linkerd/linkerd2/issues/3258#issuecomment-526377964.

**Comment by sannimichaelse** [comment]
> That's correct. Ideally, there should be a before/after comparison on the Linkerd Prometheus' memory and CPU usage, to justify this change. You can take a look at how I did it previously at [#3258 (comment)](https://github.com/linkerd/linkerd2/issues/3258#issuecomment-526377964).

Alright will compare the changes

**Comment by sannimichaelse** [comment]
Hi @ihcsim , do I need to install any package to be able to run the commands for the before and after comparison on the Linkerd Prometheus memory and CPU usage?

 following this #3258 saw the commands used 

```k -n emojivoto get deploy```  

```k -n linkerd top po```

**Comment by ihcsim** [comment]
@sannimichaelse `k` is just an alias for `kubectl` in `zsh`. The changes needed to fix this issue should be very small. Most of the work lies in the before/after CPU and memory usage comparison. If you put together a test plan, I can help you review it. See https://github.com/linkerd/linkerd2/pull/3541#issuecomment-564717323 for an example.

**Comment by srv-twry** [comment]
I'll take this one since it doesn't look like @sannimichaelse is actively working on this. Plus, this is somewhat related to the "Scale Testing" gsoc idea.

**Comment by srv-twry** [comment]
@ihcsim Is it okay if I try to use your setup described at [#3692(comment)](https://github.com/linkerd/linkerd2/issues/3692#issuecomment-564695766) but with a reduced number of deployments of `emojivoto`? May be 10 namespaces each containing 5 replicas of each. For a total of 200(10 * 5 * 4) pods.

AKS won't allow me to create 30 worker nodes because of subscription limits. 
I got this message when I tried creating more than 5 nodes: `The maximum node count you can select is 5 due to the remaining quota in the selected subscription (0 cores).`

**Comment by ihcsim** [comment]
@srv-twry Works for me. Important thing is to keep it running for a while (2+ hours?), and see the trend. Thanks!

**Comment by ihcsim** [comment]
@srv-twry Also fwiw, you will notice that the default [emojivoto.yaml](https://run.linkerd.io/emojivoto.yml) defines CPU resource requests for all the deployments, and most of the time your node are heavily underutilized. Feel free to remove those requests to allow more pods to be scheduled onto your nodes.

**Comment by srv-twry** [comment]
> Works for me. Important thing is to keep it running for a while (2+ hours?), and see the trend. Thanks!

Okay thanks :slightly_smiling_face: 

> Feel free to remove those resource requests to allow more pods to be scheduled onto your nodes.

Yes. I am actually using the scripts from your [gist](https://gist.github.com/ihcsim/4b89e43929d98ffd65658356e761f868). I saw that you've removed the requests.
